### PR TITLE
feat: normalize environment config

### PR DESCRIPTION
**Pull request by 9aoy**
## Summary

normalize environment config (environment config will merge with base rsbuild config), and support get normalize environment config via `rsbuild.getNormalizedConfig({ environment })`

resolve: https://github.com/web-infra-dev/rsbuild/issues/2632

## Checklist

- [ ] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

File: packages/core/src/createContext.ts [hunk 1]
```diff
@@ -56,6 +58,22 @@ async function createContextByConfig(
   };
 }
 
+export function updateEnvironmentContext(
+  context: RsbuildContext,
+  configs: EnvironmentConfig[],
+) {
+  context.environments = configs.map((config) => ({
```
Comment (chenjiahan) [inline packages/core/src/createContext.ts:65]: `context.environments` can be an object, which is easier to use:

```js
// object
console.log(context.environments.web.entry);

// array
console.log(context.environments.find(item => item.name === 'web')?.entry);
```

File: packages/shared/src/types/config/index.ts [hunk 2]
```diff
@@ -107,10 +107,15 @@ export type NormalizedConfig = DeepReadonly<{
   provider?: unknown;
   _privateMeta?: RsbuildConfigMeta;
   environments?: {
-    [name: string]: EnvironmentConfig;
+    [name: string]: DeepReadonly<EnvironmentOption>;
   };
 }>;
 
+/** The normalized Rsbuild environment config. */
+export type EnvironmentConfig = Omit<NormalizedConfig, 'environments'> & {
```
Comment (chenjiahan) [inline packages/shared/src/types/config/index.ts:115]: `EnvironmentConfig` and `EnvironmentOption` is confusing.

I prefer using `EnvironmentConfig` and `NormalizedEnvironmentConfig`.

File: packages/core/src/provider/initConfigs.ts [hunk 3]
```diff
@@ -54,6 +86,12 @@ export async function initRsbuildConfig({
   context.normalizedConfig = normalizeConfig(context.config);
   updateContextByNormalizedConfig(context, context.normalizedConfig);
 
+  const environments = getEnvironmentsConfig(context.normalizedConfig);
+
+  context.environmentConfigs = environments;
```
Comment (chenjiahan) [inline packages/core/src/provider/initConfigs.ts:91]: Can we just use `context.normalizedConfig.environments.[name]`?

So we do not need to add a new `context.environmentConfigs` property.

File: packages/core/src/createContext.ts (hunk 1 above)
Comment (9aoy) [reply packages/core/src/createContext.ts]: done

File: packages/shared/src/types/config/index.ts (hunk 2 above)
Comment (9aoy) [reply packages/shared/src/types/config/index.ts]: done

File: packages/core/src/provider/initConfigs.ts (hunk 3 above)
Comment (9aoy) [reply packages/core/src/provider/initConfigs.ts]: done

**Review by chenjiahan — APPROVED** [review]

**Pull request merged**